// wft-geodb-js-client/lib/model/ErrorCode.js
/*
 * GeoDB Cities API
 * The GeoDB API focuses on getting global places and regions. Easily obtain country, region, and place data for use in your apps!  <ul>   <li>Filter places by name prefix, country, location, time-zone, and even minimum population.</li>   <li>Sort places by name, country code, elevation, and population - or any combination of these.</li>    <li>Get all country regions.</li> <li>Get all places in a given country or region.</li>   <li>     Display results in multiple languages.</li> <li>RESTful API adheres to industry best-practices, including     HATEOAS-style links to facilitate paging results.   </li>    <li>Backed by cloud-based load-balanced infrastructure.</li>   <li>Data is periodically refreshed from GeoNames and WikiData.</li> </ul> <p>Notes: <ul>   <li>     Since the database is periodically updated, this may <strong>very rarely</strong> result in certain places     being marked deleted (e.g., duplicates removed). By default, endpoints returning place data will exclude     places marked deleted. However, in the unlikely event that this occurs while your app is paging through a set     of affected results - and you care about the paged results suddenly changing underneath - specify      <tt>includeDeleted=SINCE_YESTERDAY</tt> (or <tt>SINCE_LAST_WEEK</tt> if you're really paranoid!).   </li> </ul> <hr/> <h3>Useful Resources</h3> <ul>   <li>     SDKs     <ul>       <li>         <a href='https://www.npmjs.com/package/wft-geodb-angular-client'>Angular</a>,          <a href='https://github.com/wirefreethought/geodb-sample-angular-app'>Sample App</a>       </li>       <li><a href='https://github.com/wirefreethought/geodb-java-client'>Java</a></li>       <li><a href='https://www.npmjs.com/package/wft-geodb-js-client'>JavaScript</a></li>     </ul>     <li><a href='swagger.json'>Swagger Docs</a></li>     <li><a href='http://creativecommons.org/licenses/by/3.0/'>Usage License</a></li>   </li>       </ul> 
 *
 * OpenAPI spec version: 1.0.0
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 *
 * Swagger Codegen version: 2.4.30
 *
 * Do not edit the class manually.
 *
 */

(function(root, factory) {
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as an anonymous module.
    define(['ApiClient'], factory);
  } else if (typeof module === 'object' && module.exports) {
    // CommonJS-like environments that support module.exports, like Node.
    module.exports = factory(require('../ApiClient'));
  } else {
    // Browser globals (root is window)
    if (!root.GeoDb) {
      root.GeoDb = {};
    }
    root.GeoDb.ErrorCode = factory(root.GeoDb.ApiClient);
  }
}(this, function(ApiClient) {
  'use strict';

  /**
   * Enum class ErrorCode.
   * @enum {String}
   * @readonly
   */
  var exports = {
    /**
     * value: "ACCESS_DENIED"
     * @const
     */
    ACCESS_DENIED: "ACCESS_DENIED",

    /**
     * value: "ENTITY_NOT_FOUND"
     * @const
     */
    ENTITY_NOT_FOUND: "ENTITY_NOT_FOUND",

    /**
     * value: "INCOMPATIBLE"
     * @const
     */
    INCOMPATIBLE: "INCOMPATIBLE",

    /**
     * value: "PARAM_INVALID"
     * @const
     */
    PARAM_INVALID: "PARAM_INVALID",

    /**
     * value: "PARAMS_MUTUALLY_EXCLUSIVE"
     * @const
     */
    PARAMS_MUTUALLY_EXCLUSIVE: "PARAMS_MUTUALLY_EXCLUSIVE",

    /**
     * value: "REQUEST_UNPROCESSABLE"
     * @const
     */
    REQUEST_UNPROCESSABLE: "REQUEST_UNPROCESSABLE"
  };

  /**
   * Returns a <code>ErrorCode</code> enum value from a JavaScript object name.
   * @param {Object} data The plain JavaScript object containing the name of the enum value.
   * @return {module:model/ErrorCode} The enum <code>ErrorCode</code> value.
   */
  exports.constructFromObject = function(object) {
    return object;
  }

  return exports;
}));
