// wft-geodb-js-client/lib/model/PopulatedPlaceSummary.js
/*
 * GeoDB Cities API
 * The GeoDB API focuses on getting global places and regions. Easily obtain country, region, and place data for use in your apps!  <ul>   <li>Filter places by name prefix, country, location, time-zone, and even minimum population.</li>   <li>Sort places by name, country code, elevation, and population - or any combination of these.</li>    <li>Get all country regions.</li> <li>Get all places in a given country or region.</li>   <li>     Display results in multiple languages.</li> <li>RESTful API adheres to industry best-practices, including     HATEOAS-style links to facilitate paging results.   </li>    <li>Backed by cloud-based load-balanced infrastructure.</li>   <li>Data is periodically refreshed from GeoNames and WikiData.</li> </ul> <p>Notes: <ul>   <li>     Since the database is periodically updated, this may <strong>very rarely</strong> result in certain places     being marked deleted (e.g., duplicates removed). By default, endpoints returning place data will exclude     places marked deleted. However, in the unlikely event that this occurs while your app is paging through a set     of affected results - and you care about the paged results suddenly changing underneath - specify      <tt>includeDeleted=SINCE_YESTERDAY</tt> (or <tt>SINCE_LAST_WEEK</tt> if you're really paranoid!).   </li> </ul> <hr/> <h3>Useful Resources</h3> <ul>   <li>     SDKs     <ul>       <li>         <a href='https://www.npmjs.com/package/wft-geodb-angular-client'>Angular</a>,          <a href='https://github.com/wirefreethought/geodb-sample-angular-app'>Sample App</a>       </li>       <li><a href='https://github.com/wirefreethought/geodb-java-client'>Java</a></li>       <li><a href='https://www.npmjs.com/package/wft-geodb-js-client'>JavaScript</a></li>     </ul>     <li><a href='swagger.json'>Swagger Docs</a></li>     <li><a href='http://creativecommons.org/licenses/by/3.0/'>Usage License</a></li>   </li>       </ul> 
 *
 * OpenAPI spec version: 1.0.0
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 *
 * Swagger Codegen version: 2.4.30
 *
 * Do not edit the class manually.
 *
 */

(function(root, factory) {
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as an anonymous module.
    define(['ApiClient', 'model/PopulatedPlaceType'], factory);
  } else if (typeof module === 'object' && module.exports) {
    // CommonJS-like environments that support module.exports, like Node.
    module.exports = factory(require('../ApiClient'), require('./PopulatedPlaceType'));
  } else {
    // Browser globals (root is window)
    if (!root.GeoDb) {
      root.GeoDb = {};
    }
    root.GeoDb.PopulatedPlaceSummary = factory(root.GeoDb.ApiClient, root.GeoDb.PopulatedPlaceType);
  }
}(this, function(ApiClient, PopulatedPlaceType) {
  'use strict';

  /**
   * The PopulatedPlaceSummary model module.
   * @module model/PopulatedPlaceSummary
   * @version 2.0.0
   */

  /**
   * Constructs a new <code>PopulatedPlaceSummary</code>.
   * Minimal populated-place info
   * @alias module:model/PopulatedPlaceSummary
   * @class
   */
  var exports = function() {
  };

  /**
   * Constructs a <code>PopulatedPlaceSummary</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/PopulatedPlaceSummary} obj Optional instance to populate.
   * @return {module:model/PopulatedPlaceSummary} The populated <code>PopulatedPlaceSummary</code> instance.
   */
  exports.constructFromObject = function(data, obj) {
    if (data) {
      obj = obj || new exports();
      if (data.hasOwnProperty('city'))
        obj.city = ApiClient.convertToType(data['city'], 'String');
      if (data.hasOwnProperty('country'))
        obj.country = ApiClient.convertToType(data['country'], 'String');
      if (data.hasOwnProperty('countryCode'))
        obj.countryCode = ApiClient.convertToType(data['countryCode'], 'String');
      if (data.hasOwnProperty('distance'))
        obj.distance = ApiClient.convertToType(data['distance'], 'Number');
      if (data.hasOwnProperty('id'))
        obj.id = ApiClient.convertToType(data['id'], 'Number');
      if (data.hasOwnProperty('latitude'))
        obj.latitude = ApiClient.convertToType(data['latitude'], 'Number');
      if (data.hasOwnProperty('longitude'))
        obj.longitude = ApiClient.convertToType(data['longitude'], 'Number');
      if (data.hasOwnProperty('name'))
        obj.name = ApiClient.convertToType(data['name'], 'String');
      if (data.hasOwnProperty('population'))
        obj.population = ApiClient.convertToType(data['population'], 'Number');
      if (data.hasOwnProperty('region'))
        obj.region = ApiClient.convertToType(data['region'], 'String');
      if (data.hasOwnProperty('regionCode'))
        obj.regionCode = ApiClient.convertToType(data['regionCode'], 'String');
      if (data.hasOwnProperty('regionWdId'))
        obj.regionWdId = ApiClient.convertToType(data['regionWdId'], 'String');
      if (data.hasOwnProperty('type'))
        obj.type = PopulatedPlaceType.constructFromObject(data['type']);
      if (data.hasOwnProperty('wikiDataId'))
        obj.wikiDataId = ApiClient.convertToType(data['wikiDataId'], 'String');
    }
    return obj;
  }

  /**
   * The city name (legacy)
   * @member {String} city
   */
  exports.prototype.city = undefined;

  /**
   * The country name (varies by languageCode)
   * @member {String} country
   */
  exports.prototype.country = undefined;

  /**
   * The ISO-3166 country code
   * @member {String} countryCode
   */
  exports.prototype.countryCode = undefined;

  /**
   * Included if this is the result of a distance query
   * @member {Number} distance
   */
  exports.prototype.distance = undefined;

  /**
   * The place GeoDB native id
   * @member {Number} id
   */
  exports.prototype.id = undefined;

  /**
   * The place latittude (-90.0 to 90.0)
   * @member {Number} latitude
   */
  exports.prototype.latitude = undefined;

  /**
   * The place longitude (-180.0 to 180.0)
   * @member {Number} longitude
   */
  exports.prototype.longitude = undefined;

  /**
   * The place name (varies by languageCode)
   * @member {String} name
   */
  exports.prototype.name = undefined;

  /**
   * The place population
   * @member {Number} population
   */
  exports.prototype.population = undefined;

  /**
   * The region name (varies by languageCode)
   * @member {String} region
   */
  exports.prototype.region = undefined;

  /**
   * The ISO or FIPS region code
   * @member {String} regionCode
   */
  exports.prototype.regionCode = undefined;

  /**
   * The region Wikidata id
   * @member {String} regionWdId
   */
  exports.prototype.regionWdId = undefined;

  /**
   * The place type
   * @member {module:model/PopulatedPlaceType} type
   */
  exports.prototype.type = undefined;

  /**
   * The place WikiData id
   * @member {String} wikiDataId
   */
  exports.prototype.wikiDataId = undefined;


  return exports;

}));
